File: workbook/ch07/data/common/heap.c
#include <stdio.h>
#include <stdlib.h>

typedef struct {
    int *heap;
    int size;
    int capacity;
} Heap;

Heap* createHeap(int capacity);
void insertMinHeap(Heap *h, int value);
void insertMaxHeap(Heap *h, int value);
int extractMin(Heap *h);
int extractMax(Heap *h);
void heapifyUpMin(Heap *h, int index);
void heapifyUpMax(Heap *h, int index);
void heapifyDownMin(Heap *h, int index);
void heapifyDownMax(Heap *h, int index);
void swap(int *a, int *b);
void printHeap(Heap *h);
void freeHeap(Heap *h);

Heap* createHeap(int capacity) {
    Heap *h = (Heap *)malloc(sizeof(Heap));
    h->heap = (int *)malloc(capacity * sizeof(int));
    h->size = 0;
    h->capacity = capacity;
    return h;
}

void insertMinHeap(Heap *h, int value) {
    if (h->size == h->capacity) {
        printf("Heap is full\n");
        return;
    }
    h->heap[h->size] = value;
    heapifyUpMin(h, h->size);
    h->size++;
}

void insertMaxHeap(Heap *h, int value) {
    if (h->size == h->capacity) {
        printf("Heap is full\n");
        return;
    }
    h->heap[h->size] = value;
    heapifyUpMax(h, h->size);
    h->size++;
}

int extractMin(Heap *h) {
    if (h->size == 0) {
        printf("Heap is empty\n");
        return -1;
    }
    int root = h->heap[0];
    h->heap[0] = h->heap[h->size - 1];
    h->size--;
    heapifyDownMin(h, 0);
    return root;
}

int extractMax(Heap *h) {
    if (h->size == 0) {
        printf("Heap is empty\n");
        return -1;
    }
    int root = h->heap[0];
    h->heap[0] = h->heap[h->size - 1];
    h->size--;
    heapifyDownMax(h, 0);
    return root;
}

void heapifyUpMin(Heap *h, int index) {
    int parent = (index - 1) / 2;
    if (index > 0 && h->heap[index] < h->heap[parent]) {
        swap(&h->heap[index], &h->heap[parent]);
        heapifyUpMin(h, parent);
    }
}

void heapifyUpMax(Heap *h, int index) {
    int parent = (index - 1) / 2;
    if (index > 0 && h->heap[index] > h->heap[parent]) {
        swap(&h->heap[index], &h->heap[parent]);
        heapifyUpMax(h, parent);
    }
}

void heapifyDownMin(Heap *h, int index) {
    int left = 2 * index + 1;
    int right = 2 * index + 2;
    int smallest = index;

    if (left < h->size && h->heap[left] < h->heap[smallest])
        smallest = left;
    if (right < h->size && h->heap[right] < h->heap[smallest])
        smallest = right;

    if (smallest != index) {
        swap(&h->heap[index], &h->heap[smallest]);
        heapifyDownMin(h, smallest);
    }
}

void heapifyDownMax(Heap *h, int index) {
    int left = 2 * index + 1;
    int right = 2 * index + 2;
    int largest = index;

    if (left < h->size && h->heap[left] > h->heap[largest])
        largest = left;
    if (right < h->size && h->heap[right] > h->heap[largest])
        largest = right;

    if (largest != index) {
        swap(&h->heap[index], &h->heap[largest]);
        heapifyDownMax(h, largest);
    }
}

void swap(int *a, int *b) {
    int temp = *a;
    *a = *b;
    *b = temp;
}

void printHeap(Heap *h) {
    for (int i = 0; i < h->size; i++)
        printf("%d ", h->heap[i]);
    printf("\n");
}

void freeHeap(Heap *h) {
    free(h->heap);
    free(h);
}

int main() {
    Heap *minHeap = createHeap(10);
    Heap *maxHeap = createHeap(10);

    printf("Min-Heap operations:\n");
    insertMinHeap(minHeap, 5);
    insertMinHeap(minHeap, 3);
    insertMinHeap(minHeap, 8);
    insertMinHeap(minHeap, 2);
    
    printf("Min-Heap: ");
    printHeap(minHeap);

    printf("Extracted Min: %d\n", extractMin(minHeap));
    printf("Min-Heap after extraction: ");
    printHeap(minHeap);

    printf("\nMax-Heap operations:\n");
    insertMaxHeap(maxHeap, 5);
    insertMaxHeap(maxHeap, 3);
    insertMaxHeap(maxHeap, 8);
    insertMaxHeap(maxHeap, 2);

    printf("Max-Heap: ");
    printHeap(maxHeap);

    printf("Extracted Max: %d\n", extractMax(maxHeap));
    printf("Max-Heap after extraction: ");
    printHeap(maxHeap);

    freeHeap(minHeap);
    freeHeap(maxHeap);

    return 0;
}

File: workbook/ch07/data/common/heap.py

class MinHeap:
    def __init__(self):
        self.heap = []

    def _parent(self, index):
        return (index - 1) // 2

    def _left_child(self, index):
        return 2 * index + 1

    def _right_child(self, index):
        return 2 * index + 2

    def insert(self, value):
        self.heap.append(value)
        self._heapify_up(len(self.heap) - 1)

    def _heapify_up(self, index):
        parent = self._parent(index)
        if index > 0 and self.heap[index] < self.heap[parent]:
            self.heap[index], self.heap[parent] = self.heap[parent], self.heap[index]
            self._heapify_up(parent)

    def extract_min(self):
        if not self.heap:
            raise IndexError("Heap is empty")
        if len(self.heap) == 1:
            return self.heap.pop()

        root = self.heap[0]
        self.heap[0] = self.heap.pop()  # last element -> root
        self._heapify_down(0)
        return root

    def _heapify_down(self, index):
        left = self._left_child(index)
        right = self._right_child(index)
        smallest = index

        if left < len(self.heap) and self.heap[left] < self.heap[smallest]:
            smallest = left
        if right < len(self.heap) and self.heap[right] < self.heap[smallest]:
            smallest = right

        if smallest != index:
            self.heap[index], self.heap[smallest] = self.heap[smallest], self.heap[index]
            self._heapify_down(smallest)

    def peek(self):
        if not self.heap:
            raise IndexError("Heap is empty")
        return self.heap[0]

    def __str__(self):
        return str(self.heap)


class MaxHeap:
    def __init__(self):
        self.heap = []

    def _parent(self, index):
        return (index - 1) // 2

    def _left_child(self, index):
        return 2 * index + 1

    def _right_child(self, index):
        return 2 * index + 2

    def insert(self, value):
        self.heap.append(value)
        self._heapify_up(len(self.heap) - 1)

    def _heapify_up(self, index):
        parent = self._parent(index)
        if index > 0 and self.heap[index] > self.heap[parent]:
            self.heap[index], self.heap[parent] = self.heap[parent], self.heap[index]
            self._heapify_up(parent)

    def extract_max(self):
        if not self.heap:
            raise IndexError("Heap is empty")
        if len(self.heap) == 1:
            return self.heap.pop()

        root = self.heap[0]
        self.heap[0] = self.heap.pop()  # last element -> root
        self._heapify_down(0)
        return root

    def _heapify_down(self, index):
        left = self._left_child(index)
        right = self._right_child(index)
        largest = index

        if left < len(self.heap) and self.heap[left] > self.heap[largest]:
            largest = left
        if right < len(self.heap) and self.heap[right] > self.heap[largest]:
            largest = right

        if largest != index:
            self.heap[index], self.heap[largest] = self.heap[largest], self.heap[index]
            self._heapify_down(largest)

    def peek(self):
        if not self.heap:
            raise IndexError("Heap is empty")
        return self.heap[0]

    def __str__(self):
        return str(self.heap)


# example
min_heap = MinHeap()
min_heap.insert(5)
min_heap.insert(3)
min_heap.insert(8)
min_heap.insert(2)

print("Min-Heap:", min_heap)  # Output: [2, 3, 8, 5]
print("Extracted Min:", min_heap.extract_min())  # Output: 2
print("Min-Heap after extraction:", min_heap)  # Output: [3, 5, 8]

max_heap = MaxHeap()
max_heap.insert(5)
max_heap.insert(3)
max_heap.insert(8)
max_heap.insert(2)

print("Max-Heap:", max_heap)  # Output: [8, 5, 3, 2]
print("Extracted Max:", max_heap.extract_max())  # Output: 8
print("Max-Heap after extraction:", max_heap)  # Output: [5, 2, 3]
